sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqIO.java

```diff
@@ -0,0 +1,652 @@
+            return true;
+          }
+          checkpointMark.correlationIds.add(correlationId);
+        }
+        long deliveryTag = delivery.getEnvelope().getDeliveryTag();
+        checkpointMark.sessionIds.add(deliveryTag);
+        current = delivery.getBody();
+        currentTimestamp = new Instant(delivery.getProperties().getTimestamp());
+        if (currentTimestamp.isBefore(checkpointMark.oldestTimestamp)) {
+          checkpointMark.oldestTimestamp = currentTimestamp;
+        }
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+      return true;
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (channel != null) {
+        try {
+          channel.close();
+        } catch (Exception e) {
+          // ignore
+        }
+      }
+      if (connection != null) {
+        connection.close();
+      }
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to publish messages to a RabbitMQ server.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<byte[]>, PDone> {
+
+    @Nullable abstract ConnectionConfig connectionConfig();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setConnectionConfig(ConnectionConfig connectionConfig);
+      abstract Write build();
+    }
+
+    /**
+     * Define the {@link ConnectionConfig} to RabbitMQ server.
+     */
+    public Write withConnectionConfig(ConnectionConfig connectionConfig) {
+      checkArgument(connectionConfig != null, "connectionConfig can not be null");
+      return builder().setConnectionConfig(connectionConfig).build();
+    }
+
+    public Write withUri(String uri) {
+      checkArgument(uri != null, "uri can not be null");
+      return builder().setConnectionConfig(connectionConfig().withUri(uri)).build();
+    }
+
+    public Write withQueue(String queue) {
+      checkArgument(queue != null, "queue can not be null");
+      return builder().setConnectionConfig(connectionConfig().withQueue(queue)).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<byte[]> input) {
+      input.apply(ParDo.of(new WriteFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    private static class WriteFn extends DoFn<byte[], Void> {
+
+      private final Write spec;
+
+      private transient Connection connection;
+      private transient Channel channel;
+
+      public WriteFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+        ConnectionFactory connectionFactory = spec.connectionConfig().createConnectionFactory();
+        connection = connectionFactory.newConnection();
+        channel = connection.createChannel();
+        if (channel == null) {
+          throw new IOException("No RabbitMQ channel available");
+        }
+        channel.queueDeclare(spec.connectionConfig().queue(), false, false, false, null);
```

Comment:
I'm not sure what this code is trying to do. I believe RabbitMQ producers should publish to an _exchange_ rather than a queue, and they shouldn't even care whether any queues bound to this exchange currently exist:

https://www.rabbitmq.com/tutorials/tutorial-three-python.html

_The core idea in the messaging model in RabbitMQ is that the producer never sends any messages directly to a queue. Actually, quite often the producer doesn't even know if a message will be delivered to any queue at all. Instead, the producer can only send messages to an exchange...._

Reply:
Yes, forgot to merge that change. Merging.